### ISSUE TITLE
Multiple underbrace text not on the same height resp. horizontally aligned

### ISSUE THREAD

**Issue by Fabioni** [typst/typst#3784]
### Description

I think it would be nicer if the text of the `underbrace`s would be on the same height. Notice that the word "distributional" is slightly higher than "aleatoric" and "epistemic"

This is from Typst:
<img width="526" alt="image" src="https://github.com/typst/typst/assets/45362676/f517d5b1-b89a-4fb7-9244-b1b476ccb4c2">
using:
```typst
$ p(y^*|x^*, cal(D)) = integral_theta integral_mu underbrace(p(y^*|mu), "aleatoric") thick underbrace(p(mu|x^*, theta), "distributional") thick underbrace(p(theta|cal(D)), "epistemic") thick d mu thin d theta $ 
```

This is from Latex:
<img width="590" alt="image" src="https://github.com/typst/typst/assets/45362676/2cddb528-f014-4941-9b11-b89c28e60e76">

### Operating system

Web app

### Typst version

- [X] I am using the latest version of Typst

**Comment by Leedehai** [comment]
TL;DR: it looks quite contrived.

---

It appears this is the result of the different glyph renderings of the underbraces. If you look at it closely, you can notice the underbraces all look a bit differently.

In fact, if I make the middle term narrower, the middle underbrace will use _nearly_ the same glyph as the other two, and then the heights of all three _look_ aligned:

```typ
// Fig 1 (original)
$ underbrace(p(y^*|mu), "aleatoric") thick underbrace(p(mu|x^*, theta), "distributional") thick underbrace(p(theta|cal(D)), "epistemic") $ 
```
```typ
// Fig 2 (modified)
$ underbrace(p(y^*|mu), "aleatoric") thick underbrace(p(mu|x^*), "distributional") thick underbrace(p(theta|cal(D)), "epistemic") $ 
```
Fig 1: <img width="260" src="https://github.com/typst/typst/assets/18319900/9f1cb8e7-c3f1-4ae7-9bd1-92f9a08ac485"> Fig 2: <img width="260" src="https://github.com/typst/typst/assets/18319900/e8d24b44-c44d-40a0-835d-4705b2834abc">

Each term is made of three frames stacked vertically: the term itself, the underbrace, and the annotation text.

From left to right, the underbrace frames' `height()`s are:
* Fig 1: `3.4699999999999998pt`, `2.4399999999999995pt`, `3.38pt`;
* Fig 2: `3.4699999999999998pt`, `3.4699999999999998pt`, `3.38pt`.

It's not all. From left to right, the underbrace frames' `baseline()`s are:
* Fig 1: `-0.67pt`, `-1.09pt`, `-0.7000000000000001pt`;
* Fig 2: `-0.67pt`, `-0.67pt`, `-0.7000000000000001pt`.

**Comment by Leedehai** [comment]
In light of this, a sensible solution could be: instead of vertically stacking the frames freely according to their heights, we define a "quantumized" ladder of heights. For example, if the underbrace is 2pt ~ 4pt in height, we wrap the underbrace inside a frame that is 4pt in height; if the underbrace is 4pt ~ 6pt in height, we wrap the underbrace inside a frame that is 6pt in height, etc. The specific numbers are TBD.

**Comment by Enivex** [comment]
This seems to be a font issue. Here's the same thing using Stix Two Math

![image](https://github.com/typst/typst/assets/2191754/08727597-2fb9-47fe-bc6d-402d1081b025)

**Comment by Fabioni** [comment]
The font I am using is "New Computer Modern", probably a font many people use, especially in math heavy writings

**Comment by Enivex** [comment]
> The font I am using is "New Computer Modern", probably a font many people use, especially in math heavy writings 

Doesn't change the fact that the font has several issues. Though it has improved significantly over time.

**Comment by Fabioni** [comment]
It is even a used font in the Typst documentation, I think many people start from the guide for latex users: https://typst.app/docs/guides/guide-for-latex-users/.

You say this font has many problems, what alternative do you suggest?

The image from latex was just screenshotted from another paper, so I can't tell.

**Comment by Enivex** [comment]
> It is even a used font in the Typst documentation, I think many people start from the guide for latex users: https://typst.app/docs/guides/guide-for-latex-users/.
> 
> You say this font has many problems, what alternative do you suggest?

I mentioned Stix Two Math as one alternative 

> The image from latex was just screenshotted from another paper, so I can't tell.

Then it's not really a fair comparison, because it's likely the Type 1 font Computer Modern rendered through pdftex. A proper one would involve using New Computer Modern Math  through the unicode-math package with luatex or xetex.

**Comment by Leedehai** [comment]
Though it's a font issue, I still wish to fix it in Typst by changing how the `typst::math::underover::stack()` stacks up the frames - for example, we can add a param to this function that tell it whether the frame heights should be "quantumized" as mentioned above. Thoughts?

**Comment by Fabioni** [comment]
> Then it's not really a fair comparison

@Enivex, it is not about comparing against Latex, it is just about making Typst a good product. The screenshot from Latex was just to explain how I think the formula should look like. Next time I give you a hand made drawing if you wish...

**Comment by Enivex** [comment]
> > Then it's not really a fair comparison
> 
> @Enivex, it is not about comparing against Latex, it is just about making Typst a good product. The screenshot from Latex was just to explain how I think the formula should look like. Next time I give you a hand made drawing if you wish...

There was never any confusion as to what you meant. I don't think it looks great either.

My point is simply that we have to separate typst issues from font issues that have to be fixed upstream.

I've gone through the effort of comparing them:

LuaTeX + NCMM

<img width="1489" alt="luatex_ncm" src="https://github.com/typst/typst/assets/2191754/97047d47-955b-42eb-86ff-258c30b4a240">

LuaTeX + STM

<img width="1571" alt="luatex_stix" src="https://github.com/typst/typst/assets/2191754/d01b4d39-d4cb-4415-a191-e5b986d82184">

Typst + NCMM
<img width="1578" alt="typst_ncm" src="https://github.com/typst/typst/assets/2191754/b7d1c925-18e1-41f8-a1bd-7dcb6d355242">

Typst + STM
![typst_stix](https://github.com/typst/typst/assets/2191754/8b18aa38-f8e6-402f-bd81-7759dd4b1d36)

As you can see, NCMM is wonky in Typst, but it's also wonky to a lesser extent in LuaTeX. In comparison, Stix is only slightly off in Typst, and seemingly perfect in LuaTeX.

It may be a bit from column A and some from column B.

**Comment by mkorje** [comment]
I think implementing #6649 will make the Typst output match LuaTeX. (Whether that should be considered "solving" this though is another discussion...)
